### ISSUE TITLE
Allow to extend Series in types

### ISSUE THREAD

**Issue by zefirka** [leeoniya/uPlot#441]
Hi @leeoniya! I've noticed that there are a lot of uPlot plugins that I wrote depends on solution where I just put some property into Series at initialization time and then I just read it in plugin or callback. For example I'm putting into series information about series unit type for correct formatting in units plugin, or putting into series reference to original series data to show it in tooltip after data transformation along with transformed series values e.g. after normalization. 

So I thought it would be convenient to allow somehow to extend `Series` type. I can suggest to make Series type generic: 

```ts
type Series<T = {}> = {...} & T
```

But it immediately makes to turn all series containing types in generics too, which is pretty much annoying to handle:

```ts
declare class uPlot<T = {}> {
	constructor(
		opts: Options<T>,
		data?: Data,
		targ?: HTMLElement | ((self: uPlot, init: Function) => void)
	);
	...
```

Other way is to do: `class uPlot<S extends Series>`, then we'll not touch `Series`. This approach too requires to update all `Series`-containing interfaces, but it's a little bit easier to handle. 

What do you think?

**Comment by leeoniya** [comment]
`Series`, like many of uPlot's types, are interfaces, which can be extended without problems:

![image](https://user-images.githubusercontent.com/43234/106145949-ba360800-613b-11eb-91b6-08dbcd81df61.png)

**Comment by zefirka** [comment]
Sure, but in your example uPlot itself doesn't know about `MySeries` and all inner callbacks and interfaces will alarm you about type inconsistency when you'll try write some plugin which reads `series.foo` or use some callback. You can see [simplified example here](https://www.typescriptlang.org/play?#code/CYUwxgNghgTiAEkoGdnwK4AUIHsAu8A3gFACQA9AFSXwDuAFiAHbwAOKyAlkwObxTwAZuiZg8nHC0E4Y8AAJ5YPADSIoECPG6c8ACgCU-QXhCyoeRWHrc+yEBEEA6GDnzw8Od43gARAPIAsvCU5GRgksh4MOhiMrpkpDiseMgAXPB+yRJMyMoJwOZQAPzpAIIQnDxMIMA+hXmkpIowPCXwABIAKgEAMgCiECAAtswEAD7wurp2DulYuHiq2njpAGIiYtmGALwAfPAAbjicwPpk+gDcxMTwt-B2MJwgafAAyqZPyADaALpXAL7XEAAD1YMgIeAAnqwEOVKtVaoV4Nt4F8yMCAGrqdDPdJMdBDABGpl+DUc5MhWIgOJeunxRNM8Am+I0+l+pOIP2uxBBYJgBG4JhggigYAQmXEESIZAen3S70ez1+AKBoPBWiYQpFYreH2e0ru-DxBOJMCuhrBguQbV06Dm2HwS2AwONDJgO32RxOKuIgtM2oQAUhCs+8BBJiYwDQIf1JENhNdpp94RyBFluPgQZjaBRcbuUHSAEYAAx5ePpABMAE4y3dLZraegncCPQbDXcqDQAJJ0HDoCDAeAVADWCA88B4IAI4RgcDE7mhCEYcGCoXbdxTkXuyIwjnT3xOwJ+8Bu68QERwg0cuB400chMu8HI5HgACEAKqdACE1zPT+o8ClJGC4wvAoCCMs2RgTgzyar2MDDqoxJgFA6B2PAOCCBgDoEMAMHIHBw5MDgtD8ISfYEFmepoCEp7rpuaaFukVGKjmu77l8h7Hr+Z6dhkTAQJCIHgCgBBDFAQmMBArCrnR7YMfcFY7uge7UZxzrHigmbBtR5pnpul4gNeOC3sgFb3pccmGnJgKAsQClJCk6QStkbF5rc+7pF8+4-HkdkKegO7VKR8z4Lojm5KiXyFr5qIxT8lnEEAA).

Certainly many corners can be circumvented by typecasts a but I think that developers try to avoid it. If we extend all `Series` containing typings may come more complex, but developer experience will significantly be improved.

**Comment by zefirka** [comment]
I've made a little research and came to this:

 - [Example with extending `<S = {}>`](https://www.typescriptlang.org/play?#code/PQKhAIFEA8BcFMB2BnAlge0eA7q2ALcAcyXgCdUBjZcAHgGVwBecAbwF8A+ccEYAKH4ATeJQA2AQzLxw4ichoBXAApj0sBszZc2-AJCVMyWGUWVY6MgAp9e9AAdYyAFzgA8o4woGnADS2hCVgJAH5XAEExVCJEeCEAESCJfz09YLIiMPAACQAVAFkAGUgxeABbJFhwAB9wKytkeDEAM1cVNQ16P3BURDxXADFFRHMvAEpmbgA3dFQhMf0xgG5BHh5GingXcHpyVC2fAG0AXRX2QXhoe0sq2ABPexlI6NiEpK1D-WgANQkxRS2rkQijKACNyCcUgA6GF3X7-QF1YFg8g1cDAsRiMYnSH8Y6CfiXa5kKq9BBkZoSSgyDywLzITQsDjcVj6Db7ba7TYMronM4XK43cD3R47PYHRhMnRM-hrcASIEg8FkFZy65k5BZKyKNqqdQ+Xw9ITQRUosgTJjTWZCM7gABkOxW-DJ5Ep1PA+TuXI5ujloNNyv5-EMKCqDicrlp9Nonu9W24Mrl7MRh1YCvAAEYAAyG-3gABMWfYx385344eQUOTyEOWeOUPViCcWm1htQFpZsrWiir4prqHroP47CdIeM4EUWli2Aneo0sb7nCsFcNh0OGZL4HXx2Oy0EPertcH4H4QA)
- [Example with extending `<S extends Series = Series>`](https://www.typescriptlang.org/play?ssl=1&ssc=1&pln=1&pc=72#code/PQKhAIGUBcCcEMCWBzAFtAZge1gd3rACbgCmAHtCQHYDOiWV4ui0q4APJCbIiTaRWqF+XHn3ABeKN140AfOHAhgAKBWESAYwA2BEuB3wa-AK4AFbVmicBlKsOlj54AN4qAkJoY04JzdBwACg93LAAHaBoALnAAeQj6Wk45ABoQwnhoeAB+GIBBbRQqEkIAEUz4NPd3LNhkXPAACQAVAFkAGQBRbRIAW2pocAAfcEDAmhJtDBjzS2tIVPBEKhYYgDETKn9EgEpJBQA3LERCHY8dgG41RUUJpxjIAG0AXSuAXzVyMJxB6ABPML6ApFErlLKScCPDxkABq8G0Jj4MSoJl6ACNuC8qgA6XF-OEIpGjFHo7jDcAo7TaHYvLEqZ5qFRfH5LKiUWAYeCafTxaCJGg2ch2ByiWQQ0V8BRudx3WQPF7vT5kb6wX4A-QSgWQWxCEQycRSTUKKRuG7geDI1EY2BXM3fZaRBqBEwzCxWZIpJaEMiW0mwPYSQ7HQiKlQO7ic7ngVp-TU6+x6pyuFRmtG+62hry0QbhSIxXn89gxo0Q003WVEx4uC3gACMAAZPWnwAAmetvZ5pD4qXM0bEVmiPevPbH2tn8KTOz2IANSlM3Ez9-WDxAjtEqN5XFRZnzgEwQ4q4Pdu6zF5dyQK9z2PR61zuQu-PS5qRcDodr8AqIA)

 
As you can see example `<S extends Series = Series>` doesn't work with straightforward approach because of `Series` should know about type which extends itself. We can use some new types first contains only fields in `Series` that doesn't require extension and second type to combine all other fields. And define `Series` as a union of these types. [You can look at example here.](https://www.typescriptlang.org/play?ssl=1&ssc=1&pln=46&pc=1#code/PQKhAIBUE8AcFMDO4BO8BmBDAxgFwPYoCWAdgObjqHjwAeu8JiR+J4A7kbgBbgA8AZRr1GAE2QD4xJOAC84SdMQA+cOBDAAUJtHxsAG0xpwBzImQBXAAr78uPgBFhDEuIVSiSB5lyY57pW9fVQBvTQBIbFZEXBQLPEIACgjw-FhcRAAucAB5dJYmR2UAGhTRH0wAfmyAQX0iMhJ4USDMUvDw3xQyavAACUgAWQAZAFF9eABbRlxwAB9wRMTEeH10bOtbewcS8FIubIAxCxI8AoBKOVUAN3wiUXOI84BubTU1FaVsxU9EIoBtAC6rwAvto6LBCLNcHB4OA6g0mi0Kv5-hFaAA1TD6CxIbIkCyTABGUiB7QAdJToFicXjFgTiVJ5uACfp9OcgWTNIDtJoIVC9iQGCgsNg4XlcAU-k46C43D8vCj5ArEK1QhFPr9vh4kADgZowXzaJCUNDYQFfq1-GF3uBMPjCSSUKCjSazQgLUg+phXBMUNLnGIJDrVUrPaHgtaIpDSBleokLBsbHZBCGirt7rQHYyUJdZDc7qJQa9+abwDCPSrHIHXMHAmGVWr-E4AGTh72+qTS5SvTSxqSiuGDaCNlGyoPhq0295E7NOl2aKJMWZpDLZCVSvjD0eR+TTj4h7L-EL28AARgADMVwLPwAAmC8gwGlQ2rxDkzVIf4XwHkmNC5B5ATa8iDzdVbQsD8Q3+IhfyJA1eyXGJwAsfwmnYFDk3sbcQzVRI32vf5-jPZ9wGIwFARebRIM-RBvzg8BNCAA) But this approach will require types refactoring. 

 I can handle both options (example 1 or example 3), but I think that example 3 is preferable.

**Comment by leeoniya** [comment]
what's the benefit of splitting the data and the handlers instead of using `Partial<Series>` or `Pick<Series, ...>`?

also, if we go this route, i would want all the top-level major interfaces to be extendable. at least Scale, Axis, Series, Band. not sure if making Options itself extendable would make this whole thing even more twisted, but that would be good as well.

if you wanna PR it, i'll try it out and see how it goes.

**Comment by zefirka** [comment]
I'm not sure that `Partial<Series>` will work well. I think `Pick` or splitting interfaces is more explicit and reusable, but not insisting for splitting.

>also, if we go this route, i would want all the top-level major interfaces to be extendable. at least Scale, Axis, Series, Band. not sure if making Options itself extendable would make this whole thing even more twisted, but that would be good as well.

I was experimenting with types and think that we'll have to give the opportunity to reach out extended Series more or less everywhere (options - for callbacks, hooks - for plugins). I don't think that there is a reason to make Options extendable in sense of extendable Series, but in my experiments Options necessarily becomes `Options<S extends SeriesData = SeriesData>` to pass `S` to `Series`, `Scales` and all other interfaces where callbacks with `self: uPlot` argument exists. 

>if you wanna PR it, i'll try it out and see how it goes.

Great! The job is mostly routine, and if you'll grant permission I can make PR cuz I've already started experimenting and found that the result is worth it. It's really very nice when you as developer once say that you want extend uPlot series and in all your plugins and callbacks you can reach all properties in extended series.

**Comment by leeoniya** [comment]
let's leave them merged and rely on `Pick`, i don't want to bloat the exports for now.

i'm still a bit nervous about this changeset size/complexity, so don't be offended if this doesn't get merged. worst case scenario is you'll just use your own typings defs -- a nice benefit of having only external types and not generated from the source :)

**Comment by leeoniya** [comment]
someone also mentioned this, if you want to consider this route:

> there is also module augmentation, that allows you to add to an extend interfaces used by modules https://www.typescriptlang.org/docs/handbook/declaration-merging.html#module-augmentation
> (material ui promotes this as how to customise the theme  https://material-ui.com/guides/typescript/#customization-of-theme )

**Comment by zefirka** [comment]
Hi @leeoniya! I've made some experiments, and you know, once you started to extending `Series` it inevitably brings a huge amount of changes in typing if you want to work well done. So, I'm not pretty sure that there is a really necessity to bring that to uPlot, at least now. And I decided to use `.d.ts` no my side. 

But I prefer not close this issue, because I think in time a lot of people may go through this way and maybe there will be some solution someday.
